### PR TITLE
Port grouping redo/v7

### PR DESCRIPTION
**Pull request by inashivb**
[NOT ready for review bc it's ugly so far, PASSES all unittests now! 🎉  ~FAILS 22 "bad" rule tests in s-v~ They're just a result of extra notices i've put all around 🤦🏽  ]
Pre final state of port grouping revamping effort.

Previous PR: https://github.com/OISF/suricata/pull/10434
Changes since v6:
- fixed sgh updation in the interval tree
- some optimizations
- fixed boundary checks

### CODE REVIEW

File: src/detect-engine-build.c [hunk 1]
```diff
@@ -1134,10 +1136,17 @@ static DetectPort *RulesGroupByPorts(DetectEngineCtx *de_ctx, uint8_t ipproto, u
     /* step 1: create a hash of 'DetectPort' objects based on all the
      *         rules. Each object will have a SGH with the sigs added
      *         that belong to the SGH. */
+    if (direction == SIG_FLAG_TOSERVER)
+        SCLogNotice(">>>>>>>>>>>>>> TOSERVER, Proto: %s", ipproto == IPPROTO_TCP ? "tcp" : "udp");
```
Comment (victorjulien) [inline src/detect-engine-build.c:1140]: debug to get the `test-bad-*` SV tests to pass

File: src/detect-engine-build.c [hunk 2]
```diff
@@ -1134,10 +1136,17 @@ static DetectPort *RulesGroupByPorts(DetectEngineCtx *de_ctx, uint8_t ipproto, u
...
+    else
+        SCLogNotice("<<<<<<<<<<<<<< TOCLIENT, Proto: %s", ipproto == IPPROTO_TCP ? "tcp" : "udp");
     DetectPortHashInit(de_ctx);
+    SCIntervalTree *it = SCIntervalTreeInit();
```
Comment (victorjulien) [inline src/detect-engine-build.c:1144]: error check `it`

File: src/detect-engine-build.c [hunk 3]
```diff
@@ -1198,6 +1215,61 @@ static DetectPort *RulesGroupByPorts(DetectEngineCtx *de_ctx, uint8_t ipproto, u
...
+    // Creating the interval tree
+    HashListTableBucket *htb = NULL;
+    for (htb = HashListTableGetListHead(de_ctx->dport_hash_table);
+            htb != NULL;
+            htb = HashListTableGetListNext(htb))
+    {
+        DetectPort *p = HashListTableGetListData(htb);
+        int ret = 0;
+        if ((ret = PIInsertPort(de_ctx, it, &it->tree, p)) != SC_OK) {
+            SCLogNotice("ret: %d", ret);
```
Comment (victorjulien) [inline src/detect-engine-build.c:1227]: lets but an abort here while developing things, that way you catch issues early

File: src/detect-engine-build.c [hunk 4]
```diff
@@ -1198,6 +1215,61 @@ static DetectPort *RulesGroupByPorts(DetectEngineCtx *de_ctx, uint8_t ipproto, u
...
+        port = port2 = final_unique_points[0];
+        PISearchOverlappingPortRanges(de_ctx, port, port2, &it->tree, &list);
+    }
+    for (uint16_t i = 1; i < size_unique_port_arr; i++) {
+        port = final_unique_points[i - 1];
+        port2 = final_unique_points[i];
+        PISearchOverlappingPortRanges(de_ctx, port, port2, &it->tree, &list);
+    }
+    SCFree(final_unique_points);
+    }
```
Comment (victorjulien) [inline src/detect-engine-build.c:1261]: suspicious indentation

File: src/util-interval-tree.c [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+ *
+ * \author Shivani Bhardwaj <shivani@oisf.net>
+ */
+
+#include "util-interval-tree.h"
+#include "util-validate.h"
+#include "detect-engine-siggroup.h"
+#include "detect-engine-port.h"
+
+static int PICompare(SCIntervalNode *a, SCIntervalNode *b)
```
Comment (victorjulien) [inline src/util-interval-tree.c:29]: CompareAndUpdate seems more appropriate

File: src/util-interval-tree.c [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+void SCIntervalTreeFree(DetectEngineCtx *de_ctx, SCIntervalTree *it)
+{
+    if (it) {
+        SCIntervalNodeFree(de_ctx, it);
+        SCFree(it);
+    }
+}
+
+#if 1
+int COUNT = 20;
```
Comment (victorjulien) [inline src/util-interval-tree.c:73]: style: lowercase

**Review by victorjulien — CHANGES_REQUESTED** [review]
just a few comments

File: src/util-interval-tree.c [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    SCLogNotice("Inserted [%d, %d]; %d, ROOT: [%d, %d]; %d", pi->port, pi->port2,
+            IRB_COLOR(pi, irb), root->port, root->port2, IRB_COLOR(root, irb));
+    printIT(IRB_ROOT(&it->tree), 0);
+#endif
+    return SC_OK;
+}
+
+static uint16_t cnt_overlaps = 0;
+
+static bool IsOverlap(uint16_t port, uint16_t port2, SCIntervalNode *ptr)
```
Comment (victorjulien) [inline src/util-interval-tree.c:127]: all args can be `const`

File: src/util-interval-tree.c [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        printf(" ");
+    printf("[%d, %d]: %d; p = %d; max = %d\n", node->port, node->port2, IRB_COLOR(node, irb),
+            IRB_PARENT(node, irb) ? IRB_PARENT(node, irb)->port : 0, node->max);
+
+    // Process left child
+    printIT(IRB_LEFT(node, irb), space);
+}
+#endif
+
+int PIInsertPort(DetectEngineCtx *de_ctx, SCIntervalTree *it, struct PI *head, DetectPort *p)
```
Comment (victorjulien) [inline src/util-interval-tree.c:99]: `p` can be `const?

File: src/util-interval-tree.c (hunk 5 above)
Comment (victorjulien) [inline src/util-interval-tree.c:29]: `a` const?

File: src/util-interval-tree.c (hunk 8 above)
Comment (victorjulien) [reply src/util-interval-tree.c:99]: and `de_ctx` perhaps?

File: src/util-interval-tree.c [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        DetectPortFree(de_ctx, new_port);
+    return;
+}
+
+void PISearchOverlappingPortRanges(DetectEngineCtx *de_ctx,
+        uint16_t port, uint16_t port2, struct PI *head, DetectPort **list)
+{
+    SCIntervalNode *ptr = IRB_ROOT(head);
+    SCLogNotice("Finding overlaps for the range [%d, %d)", port, port2);
+    FindOverlaps(de_ctx, port, port2, ptr, &*list);
```
Comment (victorjulien) [inline src/util-interval-tree.c:188]: just `list`?

File: src/util-interval-tree.c [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+error:
+    if (new_port != NULL)
+        DetectPortFree(de_ctx, new_port);
+    return;
+}
+
+void PISearchOverlappingPortRanges(DetectEngineCtx *de_ctx,
+        uint16_t port, uint16_t port2, struct PI *head, DetectPort **list)
+{
+    SCIntervalNode *ptr = IRB_ROOT(head);
```
Comment (victorjulien) [inline src/util-interval-tree.c:186]: in theory, ptr could be NULL here and potentially paths exist in `FindOverlaps` that lead to NULL-ptr deref. Maybe not be reachable in practice, but I sleep better knowing things are checked or explained :)

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on report_test.

ERROR: QA failed on SURI_TLPW2_single_suri_time.

ERROR: QA failed on SURI_TLPW2_autofp_suri_time.

ERROR: QA failed on SURI_TLPW2_single_alerts_cmp.

ERROR: QA failed on SURI_TLPW2_autofp_alerts_cmp.

ERROR: QA failed on SURI_TLPW1_suri_time.

ERROR: QA failed on SURI_TLPR1_alerts_cmp.

ERROR: QA failed on TREX_GENERIC_cfg_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 101 | 128 | 126.73% |
| | SURI_TLPW1_stats_chk | | |
| .uptime | 138 | 147 | 106.52% |
| .detect.alert | 152920 | 138515 | 90.58% |
| .detect.alerts_suppressed | 60272 | 55717 | 92.44% |
| | SURI_TLPR1_stats_chk | | |
| .detect.alert | 3720498 | 3395878 | 91.27% |
| .detect.alerts_suppressed | 1540959 | 1487005 | 96.5% |
| | IPS_AFP_stats_chk | | |
| .detect.alerts_suppressed | 813240 | 651240 | 80.08% |
| | TREX_GENERIC_stats_chk | | |
| .detect.alerts_suppressed | 976708 | 934036 | 95.63% | 

Pipeline 18566

**Pull request closed without merging**